docs/understanding-tracking-design/json-schema-type-casting-rules/index.md

```diff
@@ -0,0 +1,1378 @@
+}
+```
+
+- `M` &le; 2147483647
+- `N` &ge; -2147483648
+- `F` is integer
+
+</td>
+<td>
+
+`INT`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+
+- `M` &le; 9223372036854775807
+- `N` &ge; -9223372036854775808
+- `F` is integer
+
+</td>
+<td>
+
+`BIGINT`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+
+- `M` &gt; 1e38-1
+- `N` &lt; -1e38
+- `F` is integer
+
+</td>
+<td>
+
+`DECIMAL(38,0)`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+
+- `M` &lt; 1e38-1
+- `N` &gt; -1e38
+- `F` is integer
+
+</td>
+<td>
+
+`DOUBLE`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "multipleOf": F
+}
+```
+
+- `F` is integer
+
+</td>
+<td>
+
+`BIGINT`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+
+
+- `P` &le; 38, where `P` is maximum precision (total number of digits) of `M` and `N`, adjusted for scale (number of digits after the `.`) of `F`.
+- `P` Rounded up to `9`, `18` or `38`, e.g. it could only take one of those 3 values.
+
+<details>
+
+`P` = `MAX`(`M.precision` - `M.scale` + `F.scale`,  `N.precision` - `N.scale` + `F.scale`)
+
+`S` = `F.scale`
+
+</details>
+
+For example, `M=10.9999, N=-10, F=0.1` will be `DECIMAL(3,1)`. Calculation as follows:
+
+`M` is `DECIMAL(6,4)`, `N` is `DECIMAL(2,0)`, `F` is `DECIMAL(2,1)`
+
+`P` = `MAX`(6 - 4 + 1, 2 + 1) = 3, rounded up to 9
+
+`S` = 1
+
+result is `DECIMAL(9,1)`
+
+</td>
+<td>
+
+`DECIMAL(P,S)`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+        
+- `P` &gt;38, where is maximum precision (total number of digits) of `M` and `N`, adjusted for scale (number of digits after the `.`) of `F`.
+
+<details>
+
+`P` = `MAX`(`M.precision` - `M.scale` + `F.scale`,  `N.precision` - `N.scale` + `F.scale`)
+
+</details>
+
+For example, `M=10.9999, N=-1e50, F=0.1` will be `DECIMAL(3,1)`. Calculation as follows:
+
+`M` is `DECIMAL(6,4)`, `N` is `DECIMAL(2,0)`, `F` is `DECIMAL(2,1)`
+
+`P` = `MAX`(6 - 4 + 1, 50 + 1) = 51
+</td>
+<td>
+
+`DOUBLE`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number", // OR ["number", "integer"]
+    "minimum": N,
+    "maximum": M,
+    "multipleOf": F
+}
+```
+
+- `M` &lt; 1e38-1
+- `N` &gt; -1e38
+- `F` is integer
+
+</td>
+<td>
+
+`DOUBLE`
+
+</td>
+</tr>
+<tr>
+<td>
+
+```json
+{
+    "type": "number" // OR ["number", "integer"]
+}
+```
+
+</td>
+<td>
+
+`DOUBLE`
+
+</td>
+</tr>
+<tr>
+<td >
+
+```json
+{
+    "enum": [N1, I1, ...]
+}
+```
+
+- **`S` = 0**
```

Comment:
What about inlining the definitions?

* All `Nx` and `Ix` are of types `number` or `integer`
* All `Nx` and `Ix` are whole numbers (or for the one below: “None of `Nx` or `Ix` are whole numbers”)
* The maximum absolute value of `Nx` and `Ix` is below ...